### PR TITLE
Disable/Enable Close traffic-light (button) on Mac OSX

### PR DESCRIPTION
**Pull request by time4OSDev**
(This is my first PR with you guys, so, if you would give me a little leeway.)

I am proposing in this PR the ability to Disable and Enable the Close traffic light on the Mac in Avalonia windows. Currently, you guys do have an EnableClose property on the Window control, but I could not find where on earth you implement it. I did find deep in your code you are doing what I found on Microsoft's sites: EnableMenuItem Win32 calls to implement the DisableCloseButton/EnableCloseButton for Windows. I decided to dig in and find out where and how to do it on Mac.

There are those in my company that brought up the valid point, when I had 'EnableCloseButton' and 'DisableCloseButton' as the methods, and they said why not just EnableClose.
1. This may imply, as they questioned, that it would mean also the Alt-F4/Cmd-W keystrokes and the upper-left (Windows) menu close and Dock right-click -> Quit (OSX) disabling as well.

I would like to propose, if you could, take this Mac implementation (in fact, or as suggestion), to have the feature on OSX, and, I think a micro and macro level of control for the traffic lights on OSX and on Windows be just button, button | menu | keystroke.

Humbly submitted.

PS. In Sandbox is a PoC showing the Disable/Enable on Mac.
If I found how you were making it the EnableClose property in Windows (from low-level win32 call), I would have liked to replicate for OSX. I.e., as property setter, rather than method calls.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=6306) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=6306) before we can accept your contribution.<br/><br/>:x: time4OSDev [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=6306)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=6306) it.</sub>

File: samples/ControlCatalog.Desktop/Program.cs [hunk 1]
```diff
@@ -18,6 +18,10 @@ public static int Main(string[] args)
         public static AppBuilder BuildAvaloniaApp()
             => AppBuilder.Configure<App>()
                 .LogToTrace()
+                .With(new AvaloniaNativePlatformOptions
```
Comment (grokys) [inline samples/ControlCatalog.Desktop/Program.cs:21]: This is debug code I assume? Should be removed.

File: samples/Sandbox/Program.cs [hunk 2]
```diff
@@ -8,6 +8,10 @@ static void Main(string[] args)
         {
             AppBuilder.Configure<App>()
                 .UsePlatformDetect()
+                .With(new AvaloniaNativePlatformOptions
```
Comment (grokys) [inline samples/Sandbox/Program.cs:11]: Again, debug code.

File: src/Avalonia.Controls/Avalonia.Controls.csproj [hunk 3]
```diff
@@ -15,7 +15,9 @@
     <ProjectReference Include="..\Avalonia.Visuals\Avalonia.Visuals.csproj" />
     <ProjectReference Include="..\Avalonia.Styling\Avalonia.Styling.csproj" />
   </ItemGroup>
+  <!--
```
Comment (grokys) [inline src/Avalonia.Controls/Avalonia.Controls.csproj:18]: I think this needs to be uncommented.

File: src/Avalonia.Controls/WindowBase.cs [hunk 4]
```diff
@@ -126,6 +126,16 @@ public void Activate()
             PlatformImpl?.Activate();
         }
 
+        public void EnableCloseButton()
```
Comment (grokys) [inline src/Avalonia.Controls/WindowBase.cs:129]: I'm not sure these methods should be added to `WindowBase`, shouldn't they be on `Window`? Popups don't have a close button.

File: samples/Sandbox/MainWindow.axaml.cs [hunk 5]
```diff
@@ -17,5 +18,14 @@ private void InitializeComponent()
         {
             AvaloniaXamlLoader.Load(this);
         }
+        private void DisableCloseButton_OnClick(object? sender, RoutedEventArgs e)
```
Comment (grokys) [inline samples/Sandbox/MainWindow.axaml.cs:21]: More debug code.

File: samples/Sandbox/MainWindow.axaml [hunk 6]
```diff
@@ -1,4 +1,12 @@
 <Window xmlns="https://github.com/avaloniaui"
         xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
+        Height="200"
+        Width="200"
         x:Class="Sandbox.MainWindow">
+    <Grid>
+        <StackPanel Orientation="Horizontal">
```
Comment (grokys) [inline samples/Sandbox/MainWindow.axaml:7]: More debug code.

File: src/Avalonia.Native/WindowImpl.cs [hunk 7]
```diff
@@ -176,6 +176,16 @@ public void SetExtendClientAreaChromeHints(ExtendClientAreaChromeHints hints)
             _native.SetExtendClientAreaHints ((AvnExtendClientAreaChromeHints)hints);
         }
 
+        // public void EnableCloseButton()
```
Comment (grokys) [inline src/Avalonia.Native/WindowImpl.cs:179]: Commented-out code should be removed.

File: src/Avalonia.Native/WindowImplBase.cs [hunk 8]
```diff
@@ -419,6 +419,15 @@ public void SetCursor(ICursorImpl cursor)
         public Action<WindowTransparencyLevel> TransparencyLevelChanged { get; set; }
 
         public IScreenImpl Screen { get; private set; }
+        public virtual void DisableCloseButton()
```
Comment (grokys) [inline src/Avalonia.Native/WindowImplBase.cs:422]: Should be moved to `WindowImpl` I think as mentioned elsewhere.

File: src/Windows/Avalonia.Win32/PopupImpl.cs [hunk 9]
```diff
@@ -17,6 +17,14 @@ class PopupImpl : WindowImpl, IPopupImpl
         [ThreadStatic]
         private static IntPtr s_parentHandle;
 
+        public void DisableCloseButton()
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/PopupImpl.cs:20]: These can be moved from `PopupImpl` if the feature is moved from `WindowBase` to `Window`.

File: src/Avalonia.Controls/WindowBase.cs [hunk 10]
```diff
@@ -126,6 +126,16 @@ public void Activate()
...
+        {
+            PlatformImpl?.EnableCloseButton();
+        }
+
+        public void DisableCloseButton()
```
Comment (grokys) [inline src/Avalonia.Controls/WindowBase.cs:134]: Any reason you implemented these as methods and not as a property? Other similar features are generally implemented as a property.

**Review by grokys — CHANGES_REQUESTED** [review]
Gave this an initial look over, I have a few comments.

@kekekeks how do you feel about adding this feature. It's obviously not available on all platforms, but we've started adding features that aren't available everywhere if the APIs are documented as to their platform limitations.

**Comment by jmacato** [comment]
Closing this due to inactivity and unresolved merge conflicts. Please ping me if you want to work on this again.

**Pull request closed without merging**